sdv/_utils.py

```diff
@@ -200,3 +201,80 @@ def _format_invalid_values_string(invalid_values, num_values):
             return f'{invalid_values[:num_values] + extra_missing_values}'
 
     return f'{invalid_values}'
+
+
+def _find_root_tables(relationships):
+    parent_tables = {rel['parent_table_name'] for rel in relationships}
+    child_tables = {rel['child_table_name'] for rel in relationships}
+    return parent_tables - child_tables
+
+
+def _get_relationship_for_child(relationships, child_table):
+    return [rel for rel in relationships if rel['child_table_name'] == child_table]
+
+
+def _get_relationship_idx_for_parent(relationships, parent_table):
+    return [
+        idx for idx, rel in enumerate(relationships) if rel['parent_table_name'] == parent_table
+    ]
+
+
+def _get_rows_to_drop(metadata, data):
+    """Get the rows to drop to ensure referential integrity.
+
+    The logic of this function is to start at the root tables, look at invalid references
+    and then save the index of the rows to drop. Then, we looked at the relationships that
+    we didn't check and repeat the process until there are no more relationships to check.
+    This ensures that we preserve the referential integrity between all the relationships.
+
+    Args:
+        metadata (MultiTableMetadata):
+            Metadata of the datasets.
+        data (dict):
+            Dictionary that maps each table name (string) to the data for that
+            table (pandas.DataFrame).
+
+    Returns:
+        dict:
+            Dictionary with the table names as keys and the indexes of the rows to drop as values.
+    """
+    table_to_idx_to_drop = {}
+    relationships = deepcopy(metadata.relationships)
+    while relationships:
+        current_roots = _find_root_tables(relationships)
+        for root in current_roots:
+            relationship_idx = _get_relationship_idx_for_parent(relationships, root)
+            for idx in relationship_idx:
```

Comment:
Instead of iterating over indices, can we just iterate over the relationships directly?